### ISSUE TITLE
Downscaling + component output: forced to toggle scanlines on/off to fix colors

### ISSUE THREAD

**Issue by amaiorano** [ramapcsx2/gbs-control#390]
When I connect my RGB-modded NES via component to the GBS, and downscale the output using the experimental component output, it works fine, except that initially the colors are wrong. They are blueish. If I toggle scanlines mode on, then off, the colors are correct.

When I either boot the gbs, or load my "downscale" preset, the screen flashes, and the colors are messed again. It looks as though scanlines mode gets enabled, then remains that way, even though it ostensibly is disabled after that. When I manually enable it, the blueish colors are present. When I disable, colors look right.

My options:
```
/preset_ntsc.A 1988
/preset_vga_upscale.B 1984
/preset_ntsc.C 1982
/slots.bin 2304
/preset_ntsc.B 1981
/preferencesv2.txt 19
preset preference = 2
frame time lock = 0
preset slot = 66
frame lock method = 0
auto gain = 0
scanlines = 0
component output = 1
deinterlacer mode = 1
line filter = 0
peaking = 1
preferScalingRgbhv = 1
6-tap = 1
pal force60 = 0
matched = 1
step response = 1
disable external clock generator = 1
```

When I load my preset:
```
preferencesv2.txt opened
loading from preset slot B: /preset_ntsc.B
ADC offset: R:42 G:43 B:42
clock gen reset: 40500000
HTotal Adjust (skipped)
Output Format: Component

preset applied: downscale (custom) for NTSC 60Hz 

source Hz: 60.09927 new out: 60.09927 clock: 40487072 (-12928)
```

**Comment by amaiorano** [comment]
Been messing around some more with this, and here's another interesting piece of info:

If I disable SyncWatcher, the scanlines on/off toggle trick doesn't fix the colors anymore. So the colors remain blueish until I enable SyncWatcher and then toggle scanlines on/off.

**Comment by amaiorano** [comment]
Another update, I managed to hack in a fix in the code. In doPostPresetLoadSteps() around line 4015, I added calls to enable then disable scanlines right after we enable the output mode:

```
    OutputComponentOrVGA();
    enableScanlines();
    disableScanlines();
```

This fixes the issue for me. It's almost certainly the wrong thing, but interestingly, when I enable scanlines, then load a preset, the scanlines mode remains enabled (the colors are still blueish), so it looks like I didn't break anything.

**Comment by amaiorano** [comment]
> If you wanna debug further, the problem is likely to be around code that enables scanlines after applying a preset. Maybe you can find where the blue hint problem occurs. If you can isolate it, you can fix it :)

Yeah, I'll see if I can do that. If you can give me more pointers as to where to look exactly, that would be great.

Btw, it's super cool that you built in the OTA updates into gbs-control. Makes it so much easier to iterate on code changes!

**Comment by ramapcsx2** [comment]
Look for calls to doPostPresetLoadSteps() and uses of rto->applyPresetDoneStage.
Please also take a look at doPostPresetLoadSteps() itself: https://github.com/ramapcsx2/gbs-control/blob/master/gbs-control.ino#L3273
I know it's enormous, but there are clear steps it does, just with lots of cases for each preset/input/option.

You may want to printf some states, to better find out which path your setup takes.

**Comment by amaiorano** [comment]
Hi Rama,

I didn't have time to debug this, but I have run into another similar issue that might help to uncover the problem. I have a PS1 that [I modded to output 15 KHz RGBHV](https://amaiorano.io/2022/04/02/ps1-vga-mod.html). When downscaling this to 15 KHz component, the output is blueish once again. Toggling scanlines seems to correct the colors, but the image is blurrier and darker (even at max intensity).

So I tried something else. I actually have 2 gbs-cs that I've built - one for downscaling to my CRT, one for upscaling to my HDTV. With the downscaler setup exactly as above (15 kHZ mode), I connected the component output into the upscaler gbs-c. The result: the colors are still blueish on my HDTV. Now, when I change the downscaler gbs-c from 15 Khz to 480p (which I can do now, since it's not going to my CRT), the colors are correct.

That was probably unclear, so here again:


PS1 -> VGAHV (15Khz) -> gbs-c downscaler (15Khz mode, component output) -> gbs-c upscaler (1080p) -> HDTV (HDMI)

Result: Blueish screen

PS1 -> VGAHV (15Khz) -> gbs-c downscaler (480p mode, component output) -> gbs-c upscaler (1080p) -> HDTV (HDMI)

Result: Correct colors

So basically when gbs-c is downscaling and outputting in component mode, the colors are wrong. Any thoughts?

**Comment by amaiorano** [comment]
I tried to debug this issue, but it looks like I can't rely on `SerialM.print*` functions. It looks like some of the prints get lost or something. I add consecutive prints like:

```
        SerialM.println(F("AY CARAMBA! 1"));
        SerialM.print(F("RGB/HV "));
```

And in the Output screen over websocket, I see:

```
AY CARAMBA! 1<reset>
Activity detected, input: RGB
VSync: present
HSync: present
AY CARAMBA! 1preferencesv2.txt opened
loading from preset slot C: /preset_vga_upscale.C
<reset>
```

I don't see the result of calling `SerialM.print(F("RGB/HV "));`. Without being able to rely on printing like this, it becomes nearly impossible to debug this. Any idea what's going on?

**Comment by ramapcsx2** [comment]
Yea, the wifi isn't reliable enough. You'll want a serial monitor, ie USB cable :p

**Comment by amaiorano** [comment]
> Yea, the wifi isn't reliable enough. You'll want a serial monitor, ie USB cable :p

Fair enough. I didn't realize the logging over wifi was this bad! I'll use a USB cable to power and log from the gbs-control. Thanks.
